### PR TITLE
GH-8487: implement HGLM gaussian [nocheck]

### PR DESCRIPTION
**Pull request by wendycwong**
This PR fixes this issue: https://github.com/h2oai/h2o-3/issues/8487

I have separated HGLM from GLM as its own toolbox.  The only family that is supported now is Gaussian.  I still need to do the following:

1. client tests (java, python/R) to make sure model metrics are passed;
2. client tests to make sure model summary, scoring history and coefficient tables are passed;
3. check and make sure we use the correct formula to estimate the residual noise variance, refer to the doc.
4. check and make sure we choose one of the likelihood methods.  I implemented two.  Refer to the doc.
[HGLM_H2O_Implementation.pdf](https://github.com/user-attachments/files/17197130/HGLM_H2O_Implementation.pdf)

### CODE REVIEW

File: h2o-py/tests/testdir_algos/glm/pyunit_benign_glm.py [hunk 1]
```diff
@@ -7,11 +7,10 @@
 
 def test_benign():
     training_data = h2o.import_file(pyunit_utils.locate("smalldata/logreg/benign.csv"))
-
+    training_data[3] = training_data[3].asfactor()
```
Comment (maurever) [inline h2o-py/tests/testdir_algos/glm/pyunit_benign_glm.py:10]: This fix should also be in a different PR.

File: h2o-py/tests/testdir_algos/glm/pyunit_link_functions_gaussian_glm.py [hunk 2]
```diff
@@ -27,6 +27,7 @@ def link_functions_gaussian():
   print("Create models with canonical link: IDENTITY")
   h2o_model = H2OGeneralizedLinearEstimator(family="gaussian", link="identity",alpha=0.5, Lambda=0)
   h2o_model.train(x=myX, y=myY, training_frame=h2o_data)
+  print(h2o_model)
```
Comment (maurever) [inline h2o-py/tests/testdir_algos/glm/pyunit_link_functions_gaussian_glm.py:30]: This fix should also be in a different PR.

File: h2o-algos/src/main/java/hex/hglm/ComputationStateHGLM.java [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        throw new IllegalArgumentException("initial_fixed_effects must be an double[] array of size "+numFixCoeff);
+
+      _beta = _parms._initial_fixed_effects;
+    } else {
+      _beta = new double[numFixCoeff];
+      _beta[_beta.length-1] = _parms.train().vec(_parms._response_column).mean();
+    }
+  }
+  
+  public double[] get_beta() { return _beta; }
```
Comment (maurever) [inline h2o-algos/src/main/java/hex/hglm/ComputationStateHGLM.java:110]: I suggest using a Camel case for every getter here in this class.
 
```suggestion
  public double[] getBeta() { return _beta; }
```

**Review by maurever — COMMENTED** [review]
Hi @wendycwong. Thanks for this big contribution.

I reviewed 80/110 files. I will continue tomorrow. It would be nice if, in this PR, you keep only HGLM-related changes. For example, implementation of HGLM can be one PR, and removing old code from GLM can be another PR. Also implementation of Python and R API can be separate PR. 

It would make the review process much easier.  Also, there would be less space for bugs.

File: h2o-algos/src/main/java/hex/hglm/HGLMModel.java [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+    public enum Method {EM}; // EM: expectation maximization
+    
+    public HGLMParameters() {
+      super();
+      _family = gaussian;
+      _em_epsilon = 1e-6;
+      _method = EM;
+    }
+
+    public GLMModel.GLMParameters.MissingValuesHandling missingValuesHandling() {
```
[the remark this answers is not in the document]
Comment (wendycwong) [reply h2o-algos/src/main/java/hex/hglm/HGLMModel.java:139]: Yes, I use the same kind of missingValuesHandling as GLM, skip or meanImputation or PlugValues

File: h2o-algos/src/main/java/hex/hglm/HGLM.java [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+      error("PlugValues", " if specified, must provide a frame with plug values in plug_values.");
+
+    if (_parms._tau_u_var_init < 0)
+      error("tau_u_var_init", "if set, must > 0.0.");
+
+    if (_parms._tau_e_var_init < 0)
+      error("tau_e_var_init", "if set, must > 0.0.");
+
+    if (_parms._seed == 0)
+      error("seed", "cannot be set to any number except zero.");
```
[the remark this answers is not in the document]
Comment (wendycwong) [reply h2o-algos/src/main/java/hex/hglm/HGLM.java:137]: Here is an example of random number generator:
image

The seed is the sequence that go into the squares (delay gates).  If that seed is all 0, the output will be 0 as well.  A history of my communications background.

File: h2o-algos/src/main/java/hex/gam/GAM.java [hunk 6]
```diff
@@ -68,7 +67,11 @@ public class GAM extends ModelBuilder<GAMModel, GAMModel.GAMParameters, GAMModel
   
   @Override
   public ModelCategory[] can_build() {
-    return new ModelCategory[]{ModelCategory.Regression};
+    return new ModelCategory[]{
```
[the remark this answers is not in the document]
Comment (wendycwong) [reply h2o-algos/src/main/java/hex/gam/GAM.java]: Done.

File: h2o-algos/src/main/java/hex/hglm/HGLMModel.java [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+    String[][] domains = new String[predictNames.length][];
+    HGLMScore gs = makeScoringTask(adaptFrm, true, j, computeMetrics);
+    gs.doAll(predictNames.length, Vec.T_NUM, gs._dinfo._adaptedFrame);
+    MetricBuilderHGLM mb = null;
+    boolean forTraining = _parms.train().getKey().equals(fr.getKey());
+    Frame rawFrame = null;
+    if (gs._computeMetrics) {
+      mb = gs._mb;
+      if (forTraining) {
+        _output._yminusxtimesz = gs._yMinusXTimesZ;
```
[the remark this answers is not in the document]
Comment (wendycwong) [reply h2o-algos/src/main/java/hex/hglm/HGLMModel.java]: I changed it myself because I need to make sure the refractor was propagated to all my codes.

File: h2o-algos/src/main/java/hex/hglm/HGLM.java [hunk 8]
```diff
@@ -0,0 +1,381 @@
...
+          model._output.setModelOutputFields(_state);
+          scoreAndUpdateModel(model, true, scTrain); // perform scoring and updating scoring history
+          if (_parms.valid() != null)
+            scoreAndUpdateModel(model, false, scValid);
+        }
+      }
+      return !stop_requested() && !converged && (_state._iter < _parms._max_iterations);
+    }
+    
+/*    public void checkPSD() {
```
[the remark this answers is not in the document]
Comment (wendycwong) [reply h2o-algos/src/main/java/hex/hglm/HGLM.java]: Done

File: h2o-algos/src/main/java/hex/hglm/HGLMModel.java [hunk 9]
```diff
@@ -0,0 +1,341 @@
...
+  @Override
+  public String toString() {
+    StringBuilder sb = new StringBuilder();
+    sb.append(super.toString());
+    sb.append(" loglikelihood: "+this._output._log_likelihood);
+    sb.append(" fixed effect coefficients: "+ Arrays.toString(this._output._beta));
+    int numLevel2 = this._output._ubeta.length;
+    for (int index=0; index<numLevel2; index++)
+      sb.append(" standard error of random effects for level 2 index " + index + ": "+this._output._tmat[index][index]);
+/*    sb.append(" standard error of residual error: "+_var_residual);
```
[the remark this answers is not in the document]
Comment (wendycwong) [reply h2o-algos/src/main/java/hex/hglm/HGLMModel.java]: Done.

File: h2o-algos/src/main/java/hex/hglm/ComputationStateHGLM.java [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+import water.util.ArrayUtils;
+import water.util.Log;
+
+import java.util.Random;
+import static hex.hglm.HGLMUtils.*;
+import static water.util.ArrayUtils.copy2DArray;
+import static water.util.ArrayUtils.gaussianVector;
+
+public class ComputationStateHGLM {
+  // the doc = document attached to https://github.com/h2oai/h2o-3/issues/8487, title HGLM_H2O_Implementation.pdf
```
[the remark this answers is not in the document]
Comment (wendycwong) [reply h2o-algos/src/main/java/hex/hglm/ComputationStateHGLM.java]: Done.

File: h2o-algos/src/main/java/hex/hglm/MetricBuilderHGLM.java [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+    if (m != null)
+      m.addModelMetrics(mm);
+    return mm;
+  }
+
+  /**
+   *
+   * This method calculates the log-likelihood as described in section II.V of the doc.
+   */
+  public static double calHGLMllg(long nobs, double[][] tmat, double varResidual, double[][][] zjTTimesZj,
```
[the remark this answers is not in the document]
Comment (wendycwong) [reply h2o-algos/src/main/java/hex/hglm/MetricBuilderHGLM.java]: I manually changed it because the change need to be propagated to all other calls of the function.

File: h2o-algos/src/main/java/hex/modelselection/ModelSelectionModel.java [hunk 12]
```diff
@@ -48,7 +48,7 @@ protected double[] score0(double[] data, double[] preds) {
 
     @Override
     public Frame score(Frame fr, String destination_key, Job j, boolean computeMetrics, CFuncRef customMetricFunc) {
-        throw new UnsupportedOperationException("AnovaGLM does not support scoring on data.  It only provide " +
+        throw new UnsupportedOperationException("ModelSelection does not support scoring on data.  It only provide " +
```
[the remark this answers is not in the document]
Comment (wendycwong) [reply h2o-algos/src/main/java/hex/modelselection/ModelSelectionModel.java]: I removed it.

File: h2o-algos/src/main/java/hex/schemas/HGLMV3.java [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+            , level = Level.expert)
+    public double em_epsilon;
+
+    @API(help="if true, will allow random component to the GLM coefficients.", direction=Direction.INPUT, gridable=true)
+    public boolean random_intercept;
+    
+    @API(help="group_column is the column that is categorical and used to generate the groups in HGLM", gridable=true)
+    public String group_column;
+
+    @API(help="if true, add gaussian noise with variance specified in parms._tau_e_var_init.",
```
[the remark this answers is not in the document]
Comment (wendycwong) [reply h2o-algos/src/main/java/hex/schemas/HGLMV3.java]: I changed if to If.

File: h2o-algos/src/test/java/hex/generic/GenericModelTest.java [hunk 14]
```diff
@@ -707,7 +707,6 @@ public void downloadable_mojo_glm() throws IOException {
             parms._train = trainingFrame._key;
             parms._distribution = AUTO;
             parms._response_column = trainingFrame._names[1];
-            parms._rand_family = new GLMModel.GLMParameters.Family[] {GLMModel.GLMParameters.Family.AUTO};
```
[the remark this answers is not in the document]
Comment (wendycwong) [reply h2o-algos/src/test/java/hex/generic/GenericModelTest.java:710]: I need to keep this one here because I rand_family is only for HGLM and not for GLM.

File: h2o-algos/src/test/java/hex/glm/GLMConstrainedTest.java [hunk 15]
```diff
@@ -917,7 +917,7 @@ public void testFindDropZeroColumns() {
     Matrix initMat = Matrix.random(11, 11);
     double[][] doubleValsOrig = (initMat.plus(initMat.transpose())).getArray();
     double[][] doubleVals = new double[doubleValsOrig.length][doubleValsOrig.length];
-    GamUtils.copy2DArray(doubleValsOrig, doubleVals);
+    copy2DArray(doubleValsOrig, doubleVals);
```
[the remark this answers is not in the document]
Comment (wendycwong) [reply h2o-algos/src/test/java/hex/glm/GLMConstrainedTest.java:920]: I moved the copy2DArray from GamUtils.java to ArrayUtils.java so that h2o-algos and h2o-core call both call it.  It needs to stay in this PR.

File: h2o-algos/src/test/java/water/test/util/PojoUtilsTest.java [hunk 16]
```diff
@@ -141,28 +141,10 @@ public void TestEquals() {
...
-    field1 = params1.getClass().getFields()[5];
-    field2 = params2.getClass().getFields()[5];
-
-    params1._rand_link = new GLMModel.GLMParameters.Link[] {GLMModel.GLMParameters.Link.identity, GLMModel.GLMParameters.Link.inverse};
-    params2._rand_link = new GLMModel.GLMParameters.Link[] {GLMModel.GLMParameters.Link.identity, GLMModel.GLMParameters.Link.inverse};
-    assert PojoUtils.equals(params1,  field1, params2,  field2);
-
-    params1._rand_link = new GLMModel.GLMParameters.Link[] {GLMModel.GLMParameters.Link.identity, GLMModel.GLMParameters.Link.inverse};
-    params2._rand_link = new GLMModel.GLMParameters.Link[] {GLMModel.GLMParameters.Link.identity};
-    assert PojoUtils.equals(params1,  field1, params2,  field2) == false;
```
[the remark this answers is not in the document]
Comment (wendycwong) [reply h2o-algos/src/test/java/water/test/util/PojoUtilsTest.java:165]: rand_link is only for HGLM toolbox and needs to be removed from GLM toolbox.  This needs to stay.  Otherwise, there will be test failures.

File: h2o-bindings/bin/custom/python/gen_gam.py [hunk 17]
```diff
@@ -15,12 +15,6 @@ def update_param(name, param):
 
 
 def class_extensions():
-    def _additional_used_columns(self, parms):
```
[the remark this answers is not in the document]
Comment (wendycwong) [reply h2o-bindings/bin/custom/python/gen_gam.py]: Yes, removed.

File: h2o-py/h2o/estimators/gam.py [hunk 18]
```diff
@@ -1595,12 +1595,6 @@ def gainslift_bins(self, gainslift_bins):
 
     Lambda = deprecated_property('Lambda', lambda_)
 
-    def _additional_used_columns(self, parms):
```
[the remark this answers is not in the document]
Comment (wendycwong) [reply h2o-py/h2o/estimators/gam.py]: Removed.

File: h2o-core/src/main/java/water/util/ArrayUtils.java [hunk 19]
```diff
@@ -16,14 +17,91 @@ public class ArrayUtils {
   public static int[] cumsum(final int[] from) {
     int arryLen = from.length;
     int[] cumsumR = new int[arryLen];
-    int result = 0;
+    int result=0;
     for (int index = 0; index < arryLen; index++) {
-      result += result+from[index];
+      result += from[index];
```
[the remark this answers is not in the document]
Comment (wendycwong) [reply h2o-core/src/main/java/water/util/ArrayUtils.java:22]: I actually need to use this method in 188 of HGLMTask.java

File: h2o-algos/src/main/java/hex/glm/GLM.java [hunk 20]
```diff
@@ -120,8 +108,10 @@ public boolean isSupervised() {
   @Override
   public ModelCategory[] can_build() {
     return new ModelCategory[]{
-      ModelCategory.Regression,
-      ModelCategory.Binomial,
+            ModelCategory.Regression,
```
[the remark this answers is not in the document]
Comment (wendycwong) [reply h2o-algos/src/main/java/hex/glm/GLM.java]: I removed it.

File: h2o-py/tests/testdir_algos/glm/pyunit_GH_6722_separate_linear_beta_gaussian.py [hunk 21]
```diff
@@ -1,7 +1,6 @@
 import h2o
 from h2o.estimators.glm import H2OGeneralizedLinearEstimator as glm
 from tests import pyunit_utils
-from tests.pyunit_utils import utils_for_glm_tests
```
[the remark this answers is not in the document]
Comment (wendycwong) [reply h2o-py/tests/testdir_algos/glm/pyunit_GH_6722_separate_linear_beta_gaussian.py:4]: I gather all utilities functions into utils_for_glm_tests (now call utils_for_glm_hglm_tests)

File: h2o-py/tests/testdir_algos/glm/pyunit_link_functions_gaussian_glm.py (hunk 2 above)
Comment (wendycwong) [reply h2o-py/tests/testdir_algos/glm/pyunit_link_functions_gaussian_glm.py]: Yes, agreed.

File: h2o-algos/src/main/java/hex/hglm/ComputationStateHGLM.java (hunk 3 above)
Comment (wendycwong) [reply h2o-algos/src/main/java/hex/hglm/ComputationStateHGLM.java]: I changed all of them.

File: h2o-py/tests/testdir_algos/glm/pyunit_benign_glm.py (hunk 1 above)
Comment (wendycwong) [reply h2o-py/tests/testdir_algos/glm/pyunit_benign_glm.py:10]: Done.

File: h2o-core/src/main/java/water/util/ArrayUtils.java [hunk 22]
```diff
@@ -1781,7 +2006,7 @@ public FrameToArray(int startCol, int endCol, long rowNum, double[][] frameConte
         _frameContent = MemoryManager.malloc8d(_rowNum, colNum);
       } else {  // make sure we are passed the correct size 2-D double array
         assert (_rowNum == frameContent.length && frameContent[0].length == colNum);
-        for (int index = 0; index < colNum; index++) { // zero fill use array
+        for (int index = 0; index < _rowNum; index++) { // zero fill use array
```
[the remark this answers is not in the document]
Comment (wendycwong) [reply h2o-core/src/main/java/water/util/ArrayUtils.java:2009]: Okay, you scared me enough that I have writeen three tests to make sure my change is correct.  I will fix tests that failed because of my change when they show up.  However, I need to first make sure that the implementation is correct.

File: h2o-algos/src/main/java/hex/hglm/HGLM.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (maurever) [reply h2o-algos/src/main/java/hex/hglm/HGLM.java:136]: Thanks for the explanation. It makes sense to me. I am surprised I have never seen this seed check in other algorithms. :thinking:  But maybe it is because I was out of work too long. ;)

File: h2o-core/src/main/java/water/util/ArrayUtils.java (hunk 22 above)
[the remark this answers is not in the document]
Comment (maurever) [reply h2o-core/src/main/java/water/util/ArrayUtils.java:2009]: Sure, I just need to warn you. :)

File: h2o-algos/src/test/java/hex/hglm/HGLMBasicTest.java [hunk 23]
```diff
@@ -0,0 +1,988 @@
...
+            .withVecTypes(T_NUM, T_NUM, T_NUM)
+            .withDataForCol(0, initUBetaT[0])
+            .withDataForCol(1, initUBetaT[1])
+            .withDataForCol(2, initUBetaT[2])
+            .build();
+    return ubetaFrame; 
+  }
+  
+  @Test
+  public void testRandomInterceptOnly() {
```
[the remark this answers is not in the document]
Comment (wendycwong) [reply h2o-algos/src/test/java/hex/hglm/HGLMBasicTest.java:693]: No, forgotten test.  I have since added test in it.

File: h2o-algos/src/test/java/hex/hglm/HGLMBasicTest.java [hunk 24]
```diff
@@ -0,0 +1,988 @@
...
+      parms._random_columns = new String[]{"x1", "x3"};
+      parms._random_intercept = true;
+      parms._group_column = "Device";
+      parms._max_iterations = 1;
+      
+      // just make sure it runs
+      HGLMModel model = new HGLM(parms).trainModel().get();
+      Scope.track_generic(model);
+      ;
+ //     ModelMetricsHGLMGaussianGaussian mmetrics = (ModelMetricsHGLMGaussianGaussian) model._output._training_metrics;
```
[the remark this answers is not in the document]
Comment (wendycwong) [reply h2o-algos/src/test/java/hex/hglm/HGLMBasicTest.java:720]: I removed the test as it does not test anything meaningful.

File: h2o-algos/src/main/java/hex/schemas/HGLMModelV3.java [hunk 25]
```diff
@@ -0,0 +1,151 @@
...
+    for (int index=0; index<tableLen; index++) {
+      tdt.set(index, 0, coeffs[index]);
+    }
+    return tdt;
+  }
+
+  public static TwoDimTable generate2DCoeffTable(String title1, String title2, double[][] coeffs, String[] coeffNames,
+                                                 String[] level2Domain) {
+    int numLevel2Index = level2Domain.length;
+    String[] coeffNamesused;
```
[the remark this answers is not in the document]
Comment (wendycwong) [reply h2o-algos/src/main/java/hex/schemas/HGLMModelV3.java]: Done

**Comment by maurever** [comment]
Hi @wendycwong. Thanks for incorporating the suggestions. There are still two HGLM tests failing. So, after all the tests pass, I can approve the PR.
